firmware/shared/code/I2C.c

```diff
@@ -0,0 +1,203 @@
+ *                 (I2C code for motor controller 2)
+ */
+
+#include "I2C.h"
+#include <stdlib.h>
+#include "stm32f4xx.h"
+
+// I2C SDA and SCL pins
+#define _I2C_SDA_GPIO  (GPIO_Pin_9) // PC9
+#define _I2C_SCL_GPIO  (GPIO_Pin_8) // PA8
+
+// I2C sendBuffer size
+#define _I2C_BUFFER_SIZE 8
+
+typedef enum{
+	I2C_STATE_SEND,    // Write
+	I2C_STATE_RECEIVE, // Read
+}I2C_state;
+
+// Global variables and their initialization
+I2C_state I2C_state_S = I2C_STATE_RECEIVE;
+
+uint8_t slave_address = 0x1;
+
+// Length of the data to be sent or received (in bytes)
+volatile uint8_t data_length = 0;
+
+// Data counter, 0 when a process finished or not started
+volatile uint8_t data_count = 0;
+
+uint8_t* sendBuffer;
+uint8_t* receiveBuffer;
+
+void I2C_init(void)
+{
+	// GPIO Init
+	RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOF, ENABLE);
+	RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOC, ENABLE);
+
+	GPIO_InitTypeDef GPIO_InitStruct;
+	GPIO_InitStruct.GPIO_Pin   = GPIO_Pin_0;
+	GPIO_InitStruct.GPIO_Mode  = GPIO_Mode_AF;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_100MHz;
+	GPIO_InitStruct.GPIO_OType = GPIO_OType_OD;
+	GPIO_InitStruct.GPIO_PuPd  = GPIO_PuPd_UP;
+	GPIO_Init(GPIOF, &GPIO_InitStruct);
+
+	GPIO_InitStruct.GPIO_Pin   = GPIO_Pin_1;
+	GPIO_InitStruct.GPIO_Mode  = GPIO_Mode_AF;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_100MHz;
+	GPIO_InitStruct.GPIO_OType = GPIO_OType_OD;
+	GPIO_InitStruct.GPIO_PuPd  = GPIO_PuPd_UP;
+	GPIO_Init(GPIOF, &GPIO_InitStruct);
+
+	// Alternative function config for SDA and SCL
+	GPIO_PinAFConfig(GPIOF, GPIO_PinSource0, GPIO_AF_I2C2);
+	GPIO_PinAFConfig(GPIOF, GPIO_PinSource1, GPIO_AF_I2C2);
+
+
+	// I2C3 Init
+	RCC_APB1PeriphClockCmd(RCC_APB1Periph_I2C2, ENABLE);
+
+	I2C_InitTypeDef I2C_Init_Struct;
+	I2C_Init_Struct.I2C_ClockSpeed          = 50000; // Any requirement for I2C clock speed?
+	I2C_Init_Struct.I2C_Ack                 = I2C_Ack_Enable;
+	I2C_Init_Struct.I2C_AcknowledgedAddress = I2C_AcknowledgedAddress_7bit;
+	I2C_Init_Struct.I2C_DutyCycle           = I2C_DutyCycle_2;
+	I2C_Init_Struct.I2C_Mode                = I2C_Mode_I2C;
+	I2C_Init_Struct.I2C_OwnAddress1         = 0x0;
+	I2C_Init(I2C1, &I2C_Init_Struct);
+
+	// Enable I2C3
+	I2C_Cmd(I2C2, ENABLE);
+
+	// Enable event interrupts & buffer interrupts
+	I2C_ITConfig(I2C2, I2C_IT_EVT, ENABLE);
+	I2C_ITConfig(I2C2, I2C_IT_BUF, ENABLE);
+
+	// NVIC Init
+	NVIC_InitTypeDef NVIC_Init_Struct;
+	NVIC_Init_Struct.NVIC_IRQChannel                   = I2C2_EV_IRQn;
+	NVIC_Init_Struct.NVIC_IRQChannelCmd                = ENABLE;
+	NVIC_Init_Struct.NVIC_IRQChannelPreemptionPriority = 0;
+	NVIC_Init_Struct.NVIC_IRQChannelSubPriority        = 0;
+	NVIC_Init(&NVIC_Init_Struct);
+
+	// Set priority for ISR and enable ISR
+	NVIC_SetPriority(I2C2_EV_IRQn, 0);
+	NVIC_EnableIRQ(I2C2_EV_IRQn);
+}
+
+bool I2C_send(I2C_channel_E channel, const uint8_t * data, const uint8_t length)
+{
+	sendBuffer = (uint8_t *)data;

```

Comment:
You need to copy the contents of `data` into a static variable in this file. `sendBuffer = (uint8_t *)data;` points `sendBuffer` to `data`. You have no control over `data` when this function returns. The contents in `data` could change.